hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -434,6 +401,107 @@ private void updateKeys(
+     * @param context pre handle context
+     * @param op token update transaction body
+     * @param originalToken original token
+     */
+    private void addRequiredSigners(
+            @NonNull PreHandleContext context,
+            @NonNull final TokenUpdateTransactionBody op,
+            @NonNull final Token originalToken)
+            throws PreCheckException {
+        if (op.hasAdminKey()) {
+            if (isValid(op.adminKey())) {
+                // if op admin key is valid we require the signature
+                context.requireKey(op.adminKeyOrThrow());
+            } else {
+                // if it's not, case is that the current admin key wants to
+                // set itself to an invalid key, hence we require only the current admin key signature
+                context.requireKey(originalToken.adminKey());
+            }
+        }
+
+        if (op.hasTreasury()) {
+            context.requireKeyOrThrow(op.treasuryOrThrow(), INVALID_ACCOUNT_ID);
+        }
+        if (op.hasAutoRenewAccount()) {
+            context.requireKeyOrThrow(op.autoRenewAccountOrThrow(), INVALID_AUTORENEW_ACCOUNT);
+        }
+
+        if (originalToken.hasAdminKey()) {
+            // if we update any of the low priority keys, we should allow either admin key
+            // or the respective low priority key to sign.
+            if (noOtherFieldThanLowPriorityKeyOrMetadataWillBeUpdated(op)) {
+                List<Key> lowPriorityKeys = new ArrayList<>();
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys, originalToken.hasWipeKey(), originalToken.wipeKey(), op.hasWipeKey());
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys, originalToken.hasKycKey(), originalToken.kycKey(), op.hasKycKey());
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys, originalToken.hasSupplyKey(), originalToken.supplyKey(), op.hasSupplyKey());
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys, originalToken.hasFreezeKey(), originalToken.freezeKey(), op.hasFreezeKey());
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys,
+                        originalToken.hasFeeScheduleKey(),
+                        originalToken.feeScheduleKey(),
+                        op.hasFeeScheduleKey());
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys, originalToken.hasPauseKey(), originalToken.pauseKey(), op.hasPauseKey());
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys,
+                        originalToken.hasMetadataKey(),
+                        originalToken.metadataKey(),
+                        op.hasMetadataKey());
+                addRequiredLowPrioritySigner(
+                        lowPriorityKeys, originalToken.hasMetadataKey(), originalToken.metadataKey(), op.hasMetadata());
```

Comment:
This update to metadata can have signature from admin or metadataKey. We need to use a threshold key in this case, like we did before.